Source/DafnyStandardLibraries/src/Std/Actions/Actions.dfy

```diff
@@ -0,0 +1,620 @@
+  //      | Done 
+  //      | Data(values: bytes)
+  //      | BadData(error: string)
+  //      | FatalError(error: string)
+  //
+  //    along with rules for what sequences of these values are valid
+  //    (e.g. once Done appears no other constructors can appear,
+  //    and perhaps if you get a FatalError the CanConsume constraints
+  //    don't even let you invoke the action again)
+  //
+  // The key point in distinguishing these semantics 
+  // is how CanConsume and CanProduce are constrained, 
+  // defining the protocol for using the action across time,
+  // depending on what inputs and outputs occur.
+  // All of the above cases are useful for precisely modeling behavior over time,
+  // and so this library provides explicity specializations for some common patterns
+  // (see Aggregators and Enumerators in particular)
+  // but allows for basically any well-founded approach.
+  @AssumeCrossModuleTermination
+  trait Action<T, R> extends GenericAction<T, R>, Validatable {
+
+    ghost var history: seq<(T, R)>
+
+    ghost predicate Valid()
+      reads this, Repr
+      ensures Valid() ==> this in Repr
+      ensures Valid() ==> CanProduce(history)
+      decreases height, 0
+
+
+    ghost predicate CanConsume(history: seq<(T, R)>, next: T)
+      requires CanProduce(history)
+      decreases height
+
+    ghost predicate CanProduce(history: seq<(T, R)>)
+      decreases height
+
+    ghost predicate Requires(t: T)
+      reads Reads(t)
+    {
+      && Valid()
+      && CanConsume(history, t)
+    }
+    ghost function Reads(t: T): set<object>
+      reads this
+      ensures this in Reads(t)
+    {
+      {this} + Repr
+    }
+    ghost function Modifies(t: T): set<object>
+      reads Reads(t)
+    {
+      Repr
+    }
+    ghost function Decreases(t: T): TerminationMetric
+      reads Reads(t)
+    {
+      NatTerminationMetric(height)
+    }
+    twostate predicate Ensures(t: T, new r: R)
+      reads Reads(t)
+    {
+      && Valid()
+      && history == old(history) + [(t, r)]
+      && fresh(Repr - old(Repr))
+    }
+
+    // Possibly optimized extensions
+
+    // Equivalent to DefaultRepeatUntil below, but may be implemented more efficiently.
+    method RepeatUntil(t: T, stop: R -> bool, ghost eventuallyStopsProof: ProducesTerminatedProof<T, R>)
+      requires Valid()
+      requires eventuallyStopsProof.Action() == this
+      requires eventuallyStopsProof.FixedInput() == t
+      requires eventuallyStopsProof.StopFn() == stop
+      requires forall i <- Consumed() :: i == t
+      reads Repr
+      modifies Repr
+      decreases Repr
+      ensures Valid()
+    //ensures history == old(history) + (n copies of t)/(n - 1 not stop values + stop)
+
+    // Convenience methods for specifications
+
+    ghost method Update(t: T, r: R)
+      reads `history
+      modifies `history
+      ensures history == old(history) + [(t, r)]
+    {
+      history := history + [(t, r)];
+    }
+
+    ghost function Consumed(): seq<T>
```

Comment:
I'm reading the names `Consumed()` and `Produced()` as if they were returned a boolean (like `Terminated()` below). A slightly longer name, but possibly better at resolving such ambiguity, is `ConsumedSoFar()` (and `ProducedSoFar()`). Hmm, maybe those have the same problem. Maybe `ConsumptionHistory` or `ConsumeHistory`? Or `ElementsConsumed` (but "elements" is not quite right)?